### ISSUE TITLE
How to keep state in bokeh server across more than one worker

### ISSUE THREAD

**Issue by volkerjaenisch** [bokeh/bokeh#7091]
Use Case: A bokeh view needs to know the IP the client requests (HTTP/WS) are coming from to make decisions.
Our View class uses the following code to determine if an incoming IP (as HTTP-Param "remote_ip"  in the request to autoload.js) is in our IP database:
```
class ProfileView(LayoutDOM):
...
    def __init__(self):
         self.set_is_privileged()

    def set_is_privileged(self):
        ip = None
        doc = curdoc()
        if doc.session_context:
            args = doc.session_context.request.arguments

            try:
                ip = args.get('remote_ip')[0]
                logger.info('%s: comes with IP %s' % (self._id, ip))
            except :
                logger.info('Exception: %s' % sys.exc_info()[0])
                logger.info('traceback: %s' % tb.format_tb(sys.exc_info()[2]))
                logger.info('%s: No ip set' % self._id)
                raise( Exception('No ip set'))
        if ip:
            ip_address = ipaddress.ip_address(ip.decode('utf8'))
            for network_entry in self._provider.privileged_ips():
                network = ipaddress.ip_network(network_entry.ip)
                if ip_address in network:
                    self._is_privileged = True
                    logger.info('%s: User is privileged.' % self._id)
                    return
        logger.info('%s: User NOT privileged.' % self._id)
        self._is_privileged = False
```
Herby self._is_privileged is the state that has to be assured over the whole session the client is connected and using the application (an interactive map and a plot).

Our network setup is not simple. We have a plone instance acting as the front portal and a Bokeh instance acting as the backend portal. 
Plone Portal:
```
server {

    location / {
        proxy_pass
        http://127.0.0.1:8083/VirtualHostBase/https/e-profile.inqbus.de:443/Plone/VirtualHostRoot/;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_http_version 1.1;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Host $host:$server_port;
        proxy_buffering off;
    }
}
```

Bokeh portal:
```
server {
    location / {
        proxy_pass http://127.0.0.1:5006;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_http_version 1.1;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Host $host:$server_port;
        proxy_buffering off;
        }
}
```
The code works as well as the nginx config. 
```
2017-10-24 00:27:18,544 INFO     3a8734ad-506b-4ffa-8ef3-991e4bd6ad46: comes with IP b'95.91.249.0'
2017-10-24 00:27:18,547 INFO     3a8734ad-506b-4ffa-8ef3-991e4bd6ad46: User is privileged.
2017-10-24 00:27:18,561 INFO     Send JS call: map.set_all_marker
2017-10-24 00:27:33,248 INFO     Send JS call: map.set_all_marker
2017-10-24 00:27:33,309 INFO     date changed to 2017-10-24
2017-10-24 00:27:33,310 INFO     Send JS call: this.aod_layer_set_date(2017-10-24)
2017-10-24 00:27:33,313 INFO     Send JS call: map.set_all_icons
2017-10-24 00:27:33,374 INFO     Send JS call: map.set_all_icons
2017-10-24 00:27:33,377 INFO     Map zoomed to level 5
```

But only if we do not use more than one bokeh worker process. And here begins the mystery.

```
2017-10-24 00:23:47,189 INFO     71378899-4b02-4a6c-adbd-ad32730f5b9e: comes with IP b'95.91.249.0'
2017-10-24 00:23:47,192 INFO     71378899-4b02-4a6c-adbd-ad32730f5b9e: User is privileged.
2017-10-24 00:23:47,205 INFO     Send JS call: map.set_all_marker
2017-10-24 00:24:01,762 INFO     traceback: ['  File "/home/graphdemoplone/workspace/eprofile/eprofile/src/eprofile/bokeh/profile/profile_layout.py", line 133, in set_is_privileged\n    ip = args.get(bcfg.IP_PARAMETER)[0]\n']247-a8e4-ae3695225d42: No ip set
2017-10-24 00:24:01,765 INFO     57c100ec-c43b-4247-a8e4-ae3695225d42: User NOT privileged.
2017-10-24 00:24:01,832 INFO     Send JS call: map.set_all_marker
2017-10-24 00:24:03,278 INFO     Send JS call: map.set_all_marker
2017-10-24 00:24:03,342 INFO     date changed to 2017-10-24
2017-10-24 00:24:03,342 INFO     Send JS call: this.aod_layer_set_date(2017-10-24)
2017-10-24 00:24:03,345 INFO     Send JS call: map.set_all_icons
2017-10-24 00:24:03,404 INFO     Send JS call: map.set_all_icons
2017-10-24 00:24:03,407 INFO     Map zoomed to level 5
```

In the case where we use more than one worker we have not repeatable issues with a second  session (self._id of the view). This means that our view class is instantiated more than once in a "user session".
This behaviour is repeatable but complex. Today I had a case where of four computers with same IP (same flat) three were accepted and the fourth not, persistently. Restarting the bokeh-Process does not change anything. Restarting nginx does not help either. As I said it is complex with to many degrees of freedom to track easily. 

Since we have no loadbalancing on the nginx (yet), all the traffic lands at the tornado instance. We have not much knowledge how tornado works. Also we have no idea what Bokeh does with sessions.

Probably we are doing something stupid and someone can point us simply to the right solution.

From a developer perspective it would be very useful to have some informations how bokeh sessions, tornado and the bokeh views interact. 

To sum up. We assume for our code that the following statements hold:
1) In a Bokeh view class e.g. derived from LayoutDOM a property like self._is_privileged keeps its state during the whole time a user is connected.
2) The state of self._is_privileged is shared across the tornado workers.

I am quite sure that 1 hold but 2 not.

Please shed some light on this topic.

**Comment by bryevdv** [comment]
@volkerjaenisch I can't comment much now but my guess is it is the same thing described here: https://github.com/bokeh/bokeh/issues/5582 

The gist: when there is more than one Bokeh server process, there is no guarantee that the initial HTTP request, and the immediate subsequent WebSocket connection end up on the same process. In such an instance IIRC you'd end up with a spurious duplicate session on the first process that the HTTP connection hit, and a real session that handles and services the real document on the process that the WebSocket hits. The spurious session would be reaped/collected after the time-out passes. I'd have to dig back into details to be sure, though. 

Having some way to make both the HTTP request and the subsequent WS upgrade hit the same process is a notion called "session affinity".  Session affinity is a hard topic.  There is nothing built into Tornado to enforce or provide it, AFAIK. And tools like Nginx only include it in their paid versions. We were able to work around the lack of session affinity enough to provide HTML request arguments, but that's it. Tho there may be some documented configurations that users have developed to make it work in free nginx. As for other proxies, I don't know even that much. 

Offhand, if you need something about the (initial HTTP) session (e.g. the IP address or anything besides the request arguments) made available to the Bokeh session, I think your best bet will be to store that information in some kind of external database (probably keyed by the session id) and then the real session that the WebSocket initiates can look it up. You could make a custom Tornado view that does this. Otherwise you will have to figure out how to implement session affinity, and I do not know how to do that.

**Comment by bryevdv** [comment]
More concretely, when you hit `app/sliders` or whatever, it goes here:

https://github.com/bokeh/bokeh/blob/master/bokeh/server/views/doc_handler.py#L21-L27

Which creates a session, and sends back HTML/JS to open a websocket connection with that session id. The WS connection ends up here:

https://github.com/bokeh/bokeh/blob/master/bokeh/server/views/ws.py

In particular, note:

https://github.com/bokeh/bokeh/blob/master/bokeh/server/views/ws.py#L96

It calls `create_session_if_needed`. "when needed" is precisely when the WS ends up on a new process that has not already created a session for that session ID previously. 

So again just to summarize, the options are:

* implement session affinity somehow, which really has to happen outside/above the level of Bokeh, or 

* store the state in an external data store (SQL db, document db, flat files..)

**Comment by volkerjaenisch** [comment]
Thank you very much for your analysis. I discussed it with my team and we will go for the second option. Lots of other projects solve the session problem in a similar fashion. At first I will try to implement https://github.com/cole/tornado-sessons at the code position you pointed me at.

**Comment by bryevdv** [comment]
As far as putting something into core Bokeh, what might make sense is to be able to provide a user defined `SessionState` object that has `save` and `load` methods or something, that (when provided) will be called by Bokeh to save/restore whatever state the developers want to keep, in whatever external store they implement with the class.

**Comment by sandrarum** [comment]
I have added a pull request which shows how we included https://github.com/cole/tornado-sessons. It is added as attribute to the classes so can be easily replaced by a custom version of the user. We are not sure how and where to provide this custom version. Maybe with some kind of Factory...
For the example I added it to bokeh core.

**Comment by bryevdv** [comment]
#9536 for 2.0 will add the ability to get the rest of the request headers/cookies, or add a hook to add anything else from the HTTP request, through to the app code execution on the WS handler, so closing this.
